### DISCUSSION TITLE
How best to work with raster data in DuckDB?

### DISCUSSION THREAD

**Discussion by markpayneatwork** [duckdb/duckdb#4411] [General]
Hi,

I am working with a large collection of satellite-images stored as rasters (matrices of data on a lon-lat grid) that I am considering storing and processing with duckdb. Each raster typically consists of 100 000 cells, and I might have, say 10 000 such rasters (one raster per day for the last 30 years). As well as storage and retrieval, I also want to do operations on them.

I am currently working with the whole system broken down into individual tuples (i.e. `cellID`, `timeID`, `value`), giving a table of around 1e9 rows. DuckDB handles this remarkibly well, but I am wondering if there are alternative structures that might be better. Some that I have considered are:
* Using the duckdb `list `data type to create arrays. But I can I perform arithmetic on these `list`s directly in duckdb (e.g. subtract one from the other) or do I have to load it into the client first?
* Going wide by creating a table with 100k columns and then adding the 10000 rows accordingly? The number of columns is fixed, but more rows come in every day.
* Blobbing the rasters and doing all the processing in the client
* Is there anything that can be done with the `struct `or `map` data types?

Any tips appreciated.

Mark

**Comment by Alex-Monahan** [comment]
Very cool! Glad things are working pretty well already!

The struct and map types are essentially making additional columns on your behalf. I'm not certain they have an advantage over explicit columns in your case. 

I think it comes down to what types of operations you need to do - do you tend to use all cells on a given day, or a subset? The extent to which your calcs only use a subset is probably the extent to which separate columns would help! Otherwise I think your current data structure or lists may work well.

Also, if you do point queries or very selective queries, an index may be helpful as well!

**Reply by markpayneatwork** [reply]
I am mainly working across time (e.g. calculate average of the rasters across a 20 year period), which is why a `list`  based approach might work. But can I do arithmetic in-situ on `list`s e.g..
```
x= [1,2,3]
y=[4,5,6]
z=y-x = [3,3,3]
```

etc..?

**Reply by Alex-Monahan** [reply]
Hmm! Maybe our list Lambda functions can support that!
https://duckdb.org/docs/sql/functions/nested
